File: TiledImageViewPrintPageRenderer.m
//
//  TiledImageViewPrintPageRenderer.m
//  iPhemeris
//
//  Created by Clifford Ribaudo on 4/16/21.
//  Copyright © 2021 Clifford Ribaudo. All rights reserved.
//
#import "TiledImageViewPrintPageRenderer.h"
#import "UIImage+Resize.h"

@implementation TiledImageViewPrintPageRenderer

-(id)init
{
    self = [super init];
    if(self) {
        _scale = 1.0;
    }
    return self;
}

-(NSInteger)numberOfPages
{
    //
    //  Determine if orientation is Portrait or Landscape and scale to fit portrait 1 wide and x down
    //  and Landscape 1 high and x wide.
    //
    _portraitOrientation = (MIN(_fullImage.size.width, _fullImage.size.height) == _fullImage.size.width) ? YES : NO;
    if(_portraitOrientation)
        _scale = self.printableRect.size.width / _fullImage.size.width;
    else
        _scale = self.printableRect.size.height / _fullImage.size.height;
    
    // If image is smaller than printable rect don't scale up.
    _scale = (_scale > 1) ? 1.0 : _scale;
    
    if(_scale < 1) {
        CGSize scaledSize = CGSizeMake(round(_fullImage.size.width * _scale), round(_fullImage.size.height * _scale));
        _scaledImage = [_fullImage imageWithSize:scaledSize];
    }
    else
        _scaledImage = _fullImage;
    
    CGFloat fullWidth = _scaledImage.size.width;
    CGFloat fullHeight = _scaledImage.size.height;
    
    NSInteger pageCnt = 0;
    while(YES)
    {
        pageCnt++;
        if(_portraitOrientation) {
            fullHeight -= round(self.printableRect.size.height);
            if(fullHeight < 0) break;
        }
        else {
            fullWidth -= round(self.printableRect.size.width);
            if(fullWidth < 0) break;
        }
    }
    return pageCnt;
}

-(void)drawPageAtIndex:(NSInteger)pageIndex inRect:(CGRect)printableRect
{
    printableRect.origin.x = round(printableRect.origin.x);
    printableRect.origin.y = round(printableRect.origin.y);
    printableRect.size.width = round(printableRect.size.width);
    printableRect.size.height = round(printableRect.size.height);
    
    CGRect cropRect;
    cropRect.origin.x = 0;
    cropRect.origin.y = 0;
    cropRect.size = printableRect.size;
    
    if(_portraitOrientation)
    {
        CGFloat remainingHeight;
        printableRect.origin.y += (printableRect.size.height * pageIndex);
        remainingHeight = _scaledImage.size.height - cropRect.origin.y;
        
        if(remainingHeight < printableRect.size.height)
            printableRect.size.height = remainingHeight;
    }
    else {
        cropRect.origin.x = (printableRect.size.width * pageIndex);
        cropRect.size.width = _scaledImage.size.width - cropRect.origin.x;
        
        if(cropRect.size.width > printableRect.size.width)
            cropRect.size.width = printableRect.size.width;
    }
    // Get the portion of the scaled image that should print on the current page
    UIImage *imageForPage = [_scaledImage subImageForRect:cropRect atScale:0.0];

    // Print that image portion on the current page
    [imageForPage drawAtPoint:printableRect.origin];
}
@end

File: TiledImageViewPrintPageRenderer.swift
//
//  TiledImageViewPrintPageRenderer.swift
//  iPhemeris
//
//  Created by Clifford Ribaudo on 12/29/24.
//  Copyright © 2024 Clifford Ribaudo. All rights reserved.
//
import Foundation
import UIKit

@objc class TiledImageViewPrintPageRenderer : UIPrintPageRenderer
{
    var scale: CGFloat = 1.0
    var portraitOrientation = true
    var scaledImage = UIImage()
    
    @objc var fullImage = UIImage()
    @objc override init() {
        super.init()
    }
    
    @objc override var numberOfPages: NSInteger
    {
        //
        //  Determine if orientation is Portrait or Landscape and scale to fit portrait 1 wide and x down
        //  and Landscape 1 high and x wide.
        //
        portraitOrientation = (min(fullImage.size.width, fullImage.size.height) == fullImage.size.width) ? true : false
        
        if(portraitOrientation) {
            scale = self.printableRect.size.width / fullImage.size.width
        } else {
            scale = self.printableRect.size.height / fullImage.size.height
        }
        // If image is smaller than printable rect don't scale up.
        scale = (scale > 1) ? 1.0 : scale
        
        scaledImage = fullImage
        if(scale < 1) {
            let scaledSize = CGSizeMake(round(fullImage.size.width * scale), round(fullImage.size.height * scale))
            scaledImage =  fullImage.imageWithSize(scaledSize)!
        }
        
        var fullWidth = scaledImage.size.width
        var fullHeight = scaledImage.size.height
        var pageCnt: NSInteger = 0
        while(true) {
            pageCnt += 1
            if(portraitOrientation) {
                fullHeight -= round(self.printableRect.size.height)
                if(fullHeight < 0) { break }
            }
            else {
                fullWidth -= round(self.printableRect.size.width)
                if(fullWidth < 0) { break }
            }
        }
        return pageCnt
    }
    
    @objc override func drawPage(at pageIndex: Int, in printableRect: CGRect)
    {
        var rectToPrint = printableRect
        rectToPrint.origin.x = round(rectToPrint.origin.x);
        rectToPrint.origin.y = round(rectToPrint.origin.y);
        rectToPrint.size.width = round(rectToPrint.size.width);
        rectToPrint.size.height = round(rectToPrint.size.height);
        
        var cropRect = CGRect(origin:CGPoint(x:0, y:0), size:rectToPrint.size)

        if(portraitOrientation) {
            cropRect.origin.y = (rectToPrint.size.height * CGFloat(pageIndex))
            cropRect.size.height = scaledImage.size.height - cropRect.origin.y
            if(cropRect.size.height > printableRect.size.height) {
                cropRect.size.height = printableRect.size.height
            }
        }
        else {
            cropRect.origin.x = (printableRect.size.width * CGFloat(pageIndex))
            cropRect.size.width = scaledImage.size.width - cropRect.origin.x
            if(cropRect.size.width > printableRect.size.width) {
                cropRect.size.width = printableRect.size.width
            }
        }
        
        // Get the portion of the scaled image that should print on the current page
        let imageForPage = scaledImage.subImageForRect(cropRect, atScale: 0.0)

        // Print that image portion on the current page
        imageForPage?.draw(at: rectToPrint.origin)
    }
}
